### ISSUE TITLE
ComboBox DataCommunicator can get stuck, leading to OutOfMemoryError

### ISSUE THREAD

**Issue by p-engl** [vaadin/flow-components#4674]
### Description

We encountered a situation where the DataCommunicator of a ComboBox can get stuck in an infinite loop which eventually leads to an OutOfMemoryError. I created a minimal example to reproduce the problem.

Some additional information about this example:
I am aware that this isolated example does not make very much sense on its own. In our application the problem occured at a place where the ComboBox was removed from its StateTree to avoid problems related to the usage of PreserveOnRefresh ( java.lang.IllegalStateException: Can't move a node from one state tree to another).
In any case I would argue that the DataCommunicator should never end up in an infinite loop.

Issue [#14435](https://github.com/vaadin/flow/issues/14435) might be related.

### Expected outcome

The DataCommunicator should not get stuck in an infinite loop.

### Minimal reproducible example

```
@Route("test")
public class TestView extends Div{

	public TestView() {
		setSizeFull();
		
		ComboBox<String> cb = new ComboBox<>();
		
		Button btn = new Button("Test");
		add(btn);
		btn.addClickListener(e -> {
			cb.setItems(new ArrayList<>());
			remove(cb);
			
			cb.getElement().removeFromTree();
			cb.setItems(new ArrayList<>(Arrays.asList("a","b","c")));
			add(cb);
		});
		add(cb);
	}
}
```

### Steps to reproduce

Go to /test and click the button.

### Environment

Vaadin version(s): 23.3.2
OS: Windows 10

### Browsers

Issue is not browser related

**Comment by TatuLund** [comment]
> cb.getElement().removeFromTree();

This should not be necessary if you have already removed the component. Thus asking what purpose this line of code serves here? I.e. what actual real life scenario it tries to simulate? Without this line of code there is no issues.

**Comment by p-engl** [comment]
We have a quite data intensive application with several views which can be created and configured by the user and which also allows multiple ways of data filtering, manipulation and visualisation. Our users expect that the state of the views are preserved even when navigating between views via browser bookmarks. The alternative to reusing views would be to have some data models which store the state of the UI so that a view can be recreated from this data model. This however leads to additional complexity and developement time because at the end of the day we just want to display the same view that was already displayed in the same browser tab.
That's why we implemented something like a "browser tab scope", i. e. views are reused within the same browser tab when navigating via UI.navigate as well as manually navigating via url manipulation or browser bookmarks. This is also the same way that our application worked in Vaadin 8. This requires removeFromTree() at certain points because views or parts of them may still be attached to an old UI instance. 

Another special "feature" in our application is that we use a custom ComponentRenderer variation which allows us to "reuse" components. Just to give you an idea: We have a grid composite as the base class of all our tables which allows you to define columns like this:
`grid.addColumn("some-column-identifier", Component.class, item -> new CheckBox(item.foo()));`
In this case a CheckBox is generated once for each item and reused when scrolling through the grid (so that the state of the component does not get lost when you scroll through the grid). This also requires removeFromTree() in our custom ComponentRenderer (to be more precise: in ComponentDataGenerator.destroyData and ComponentDataGenerator.destroyAllData), otherwise we would get an exception after browser refresh because the components displayed in the grid might still be attached to the old UI.
I am aware that this is not exactly the recommended way of rendering components in a grid. You would probably recommend to have some data object in the background from which the component is recreated each time the grid renders a cell and the data model is synchronized using some value change listeners etc. But just like above, at the end of the day I simply want to display a Component in a grid cell. The example with the checkbox is trivial, but we often have more complex components with conditional styling. 

So to summarize: We call removeFromTree() at certain points in our application because we reuse components (in the same browser tab and in grids). This is normally no problem. There were however two occasions during developement in the last weeks where, due to an unfortunate order of refreshes, something like in my initial post happened. The problem was easy to resolve during developement, but I am afraid that such a problem might also appear in a production environment.

I understand that this is no problem that appears in a "normal" Vaadin application. However I would be very grateful if there was some "bugfix" which avoids an OutOfMemoryError in the example I provided, although this problem is probably not your top priority right now.

**Comment by ugur-vaadin** [comment]
This issue was fixed by a [PR](https://github.com/vaadin/flow/pull/17231) in the Flow repo.
